src/translator/quality_estimator.cpp

```diff
@@ -0,0 +1,165 @@
+#include "quality_estimator.h"
+
+#include <iostream>
+
+#include "byte_array_util.h"
+
+namespace marian {
+namespace bergamot {
+
+QualityEstimator::QualityEstimator(const AlignedMemory& qualityEstimatorMemory) {
+  LOG(info, "[data] Loading Quality Estimator model from buffer");
+  QualityEstimator::load(qualityEstimatorMemory.begin(), qualityEstimatorMemory.size());
+}
+
+void QualityEstimator::load(const char* ptr, size_t blobSize) {
+  /* File layout:
+   * header
+   * stds array
+   * means array
+   * coefficients array
+   * intercepts array
+   */
+  ABORT_IF(blobSize < HEADER_SIZE, "Quality estimation file too small");
+  const Header& header = *reinterpret_cast<const Header*>(ptr);
+  numFeatures_ = header.numFeatures / sizeof(float);
+  ptr += sizeof(Header);
+  ABORT_IF(header.magic != BINARY_QE_MODEL_MAGIC, "Incorrect magic bytes for quality estimation file");
+  uint64_t expectedSize = sizeof(Header) + numFeatures_ * sizeof(float) * 4;  // stds, means, intercept, coef
+  ABORT_IF(expectedSize != blobSize, "QE header claims file size should be {} bytes but file is {} bytes", expectedSize,
+           blobSize);
+  const float* begin = reinterpret_cast<const float*>(ptr);
+  const int numParameters = 4;
+  stds_ = begin;
+  means_ = (begin += numFeatures_);
+  coefficients_ = (begin += numFeatures_);
+  intercept_ = (begin += numFeatures_);
+  modelParameters_.resize(numParameters * numFeatures_);
+  //[0,1,2,3]-> stds, means, coefficients, intercept
+  for (int i = 0; i < numFeatures_; i++) {
+    modelParameters_[i + 0 * numFeatures_] = stds_[i];
+    modelParameters_[i + 1 * numFeatures_] = means_[i];
+    modelParameters_[i + 2 * numFeatures_] = coefficients_[i];
+    modelParameters_[i + 3 * numFeatures_] = intercept_[i];
+  }
+}
+
+QualityEstimator::SentenceQualityEstimate QualityEstimator::mapBPEToWords(Quality quality, AnnotatedText target,
+                                                                          size_t sentenceIdx) const {
+  const ByteRange& sentenceByteRange = target.sentenceAsByteRange(sentenceIdx);
+  QualityEstimator::SentenceQualityEstimate sentenceQualityScores;
+  int bpeLen = 1;
+  int wordIdx = 0;
+  int subwordIdx = wordIdx;
+  float tmp;
+  for (float& subwordScore : quality.word) {
+    ByteRange subword = target.wordAsByteRange(sentenceIdx, subwordIdx);
+    if (subword.begin == sentenceByteRange.end) {  // EOS token
+      insertNewWord(sentenceQualityScores, subword, subwordScore, 1);
+      break;
```

Comment:
The reason for this break is that it is the only way that I found to collect the EOS token without target. text breaks since it does not have to access to it. I know that this is not ideal from a software engineering point of view so suggestions are welcome here.

Reply:
If somehow the model predicts an `<unk>` (which is of zero length on surface - text representation) at the end, this condition is true and you'll skip a token wouldn't you?